app/activeproject.cpp

```diff
@@ -465,31 +450,28 @@ void ActiveProject::setMapTheme( const QString &themeName )
 
 void ActiveProject::updateActiveLayer()
 {
+  QList< QgsMapLayer * > visibleLayers = getVisibleLayers();
+
+  if ( !mQgsProject || visibleLayers.isEmpty() )
+    return;
+
   if ( !InputUtils::layerVisible( mActiveLayer.layer(), mQgsProject ) )
   {
-    QgsMapLayer *defaultAppSettingsLayer = InputUtils::mapLayerFromName( mAppSettings.defaultLayer(), mQgsProject );
-
-    if ( InputUtils::recordingAllowed( defaultAppSettingsLayer, mQgsProject ) )
-    {
-      setActiveLayer( defaultAppSettingsLayer );
-      return;
-    }
+    QgsMapLayer *defaultLayer = InputUtils::mapLayerFromName( mAppSettings.defaultLayer(), mQgsProject );
 
-    // default layer from app settings has no recording allowed => let's try to search for a new one
-    QgsMapLayer *defaultLayer = nullptr;
-    const QMap<QString, QgsMapLayer *> layers = mQgsProject->mapLayers();
-    for ( auto it = layers.cbegin(); it != layers.cend(); ++it )
+    if ( !recordingAllowed( defaultLayer ) || !defaultLayer )
```

Comment:
Also here, `!defaultLayer` is checked within `recordingAllowed` first, so it's practically unreachable.